generative/networks/blocks/selfattention.py

```diff
@@ -95,15 +108,25 @@ def forward(self, x: torch.Tensor, context: torch.Tensor | None = None) -> torch
         key = key.view(b, kv_t, self.num_heads, c // self.num_heads).transpose(1, 2)  # (b, nh, kv_t, hs)
         value = value.view(b, kv_t, self.num_heads, c // self.num_heads).transpose(1, 2)  # (b, nh, kv_t, hs)
 
-        # manual implementation of attention
-        attention_scores = (query @ key.transpose(-2, -1)) * self.scale
+        if self.use_flash_attention:
+            query = query.contiguous()
+            key = key.contiguous()
+            value = value.contiguous()
+            y = xops.memory_efficient_attention(
+                query, key, value, attn_bias=xops.LowerTriangularMask() if self.causal else None
+            )
+
+        else:
+            # manual implementation of attention
+            attention_scores = (query @ key.transpose(-2, -1)) * self.scale
+
+            if self.causal:
+                attention_scores = attention_scores.masked_fill(self.causal_mask[:, :, :t, :kv_t] == 0, float("-inf"))
```

Comment:
These two methods give different values. `scale` isn't currently passed to `memory_efficient_attention`, nor the dropout probability, but even account for that it looks like that isn't the root of the difference. The python implementation of `memory_efficient_attnetion` looks a little different to the manual implementation here:

https://github.com/facebookresearch/xformers/blob/658ebab39545f180a6075385b3897921623d6c3b/xformers/ops/fmha/__init__.py#L142-L149

Reply:
thanks for pointing this, I fixed the problem

```
import torch

from generative.networks.blocks import SABlock

device = torch.device("cuda")
sab = SABlock(hidden_size=4, num_heads=2, dropout_rate=0.0, use_flash_attention=False)
sab = sab.to(device)
sab.eval()

with torch.no_grad():
    x = torch.randn(1, 3, 4).to(device)
    result = sab(x)

    sab.use_flash_attention = True
    result_flash = sab(x)

torch.isclose(result, result_flash)
```
returning
```
tensor([[[True, True, True, True],
         [True, True, True, True],
         [True, True, True, True]]], device='cuda:0')
```